zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/randomized/ReplayStateRandomizedPropertyTest.java

```diff
@@ -168,14 +165,10 @@ private void processingHasStoppedAndNoNewRecordsAreAddedDuringGracePeriod()
     assertThat(engineRule.hasReachedEnd())
         .describedAs("Processing has reached end of the log.")
         .isTrue();
-    final var stateBeforeGracePeriod = engineRule.collectState();
     Thread.sleep(GRACE_PERIOD);
     assertThat(engineRule.hasReachedEnd())
         .describedAs("Processing has reached end of the log.")
         .isTrue();
```

Comment:
❓ Do we still need this grace period wait time?

Reply:
That makes a lot of sense, thanks @korthout 👍 

I'm wondering what could still go wrong without waiting for another grace period:
* Before we stop the engine, we collect the last position and the state.
* After we collect the state and position, another event is processed.
* We stop the engine.
* We restart the engine in REPLAY mode.
* The engine replays more events than in the first run, at least according to our stored last position and collected state.

I'm not 100% confident that this would fail the test, happy for any further thoughts on this.

Reply:
The thing is that we pause the processing before collecting the state and position. I think, but I don't know for sure, that after pausing it's possible for the engine to wrap up processing the current command it is processing.

What I mean is that the processing state machine moves through its states in steps, each executed as a task by the actor scheduler. Pausing is also just a task executed by the actor scheduler. I believe this can be exeucted in between any of the tasks, but only takes affected at the start of each new cycle of states:
- In `tryToReadNextRecord` it checks `shouldProcessNext` which is flipped by pausing: https://github.com/camunda/camunda/blob/889db69401e94fc86d5f1dd8daf7d0ba35f458d3/zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java#L238

Note that pausing joins on it's own execution, not the completion of the cycle of processing state machine.

So, the grace period could protect against the following:
- we pause the engine (flipping the boolean)
- we collect the state
- the engine still appends some new events, changing state
- we stop the engine
- ... 

With that in mind, the strange thing is that I've already made 350 successful passes of the test without the grace period.

Reply:
Yay for async systems 😄
I think the possibility of the test still failing now without the grace period is simply super small.
There might only be a theoretical chance left now to get into this edge case, if at all possible.

Reply:
@tmetzke Let's keep the grace period in for now then. I'd prefer a deterministic slightly slower test over a faster but possibly flaky test.

We would be able to remove this grace period, if `pause` would only return once the ProcessingStateMachine is paused **and** has completed a full cycle. @camunda/zeebe-distributed-platform is that something you would be willing to add to the admin pause endpoint?